File: Java Extras/src/Snippets/TailRecursion.java
package Snippets;

public class TailRecursion {
    public static void main(String[] args) {
        System.out.println(fact(4));
        System.out.println(tail_fact(4, 1));
    }

    //regular recursion - inefficient
    static int fact(int n) {
        if (n == 1)
            return 1;
        return n * fact(n - 1);
    }

    //tail recursion - more efficient
    static int tail_fact(int n, int m) {
        if (n == 1)
            return m;
        return tail_fact(n - 1, n * m);
    }

}

File: Data Structures/src/recursions/TailRecursion.kt
package recursions

import java.math.BigInteger

val ONE: BigInteger = BigInteger.ONE
val FIFTY = 50.toBigInteger()

fun main() {
    println(tailFact(FIFTY * FIFTY))
    println(fact(FIFTY * FIFTY))

    // now regular recursion will fail
    println(tailFact(FIFTY * FIFTY * FIFTY))
    println("~".repeat(100))
    println(fact(FIFTY * FIFTY * FIFTY))
}

//regular recursion - inefficient
fun fact(n: BigInteger): BigInteger {
    return if (n == ONE) ONE else n * fact(n - ONE)
}

//tail recursion - more efficient
tailrec fun tailFact(n: BigInteger, m: BigInteger = ONE): BigInteger {
    return if (n == ONE) m else tailFact(n - ONE, n * m)
}